sopt/src/main/scala/dagr/sopt/cmdline/CommandLineParser.scala

```diff
@@ -0,0 +1,527 @@
+                                      clpBlock: Clp => Unit.type = c => Unit): Option[Clp] = {
+
+    /** A brief developer note about the precedence of printing error messages and usages.
+      *
+      * See [[parseMainAndClp]] for 1-3, for when we have options to the main tool.
+      *
+      * 4. Complain if (a) no clp name is given, or (b) it is not recognized.
+      * 5. Complain if clp options are mis-specified.
+      * 6. Print the clp help message if --help is used after a clp name.
+      * 7. Print the version if --version is used after a clp name.
+      * 8. ... execute the clp ...
+      * */
+
+    val errorMessageBuilder: mutable.StringBuilder = new StringBuilder()
+
+    val print : (String => Unit) = System.err.println
+
+    def printExtraUsage(clazz: Option[Class[_]]): Unit = {
+      extraUsage match {
+        case Some(str) => print(str)
+        case None => print(standardClpUsagePreamble(clazz))
+      }
+    }
+
+    // Try parsing the task name
+    val classToClpAnnotation = findClpClasses[Clp](packageList, omitSubClassesOf = omitSubClassesOf, includeHidden = includeHidden)
+    parseClpName(args = args, classes = classToClpAnnotation.keySet) match {
+      case Right(error) => // Case 4 (b)
+        printExtraUsage(None)
+        print(clpListUsage(classToClpAnnotation.keySet, commandLineName, withPreamble=false))
+        print(wrapError(error))
+        None
+      case Left(clazz) =>
+        /////////////////////////////////////////////////////////////////////////
+        // Parse the arguments for the Command Line Program class
+        /////////////////////////////////////////////////////////////////////////
+        val clpParser = new CommandLineProgramParser(clazz)
+        clpParser.parseAndBuild(errorMessageBuilder, args.drop(1)) match {
+          case ParseResult.Failure => // Case 5
+            printExtraUsage(clazz=Some(clazz))
+            print(clpParser.usage(withPreamble=false))
+            print(wrapError(errorMessageBuilder.toString()))
+            None
+          case ParseResult.Help => // Case 6
+            printExtraUsage(clazz=Some(clazz))
+            print(clpParser.usage(withPreamble=false))
+            None
+          case ParseResult.Version => // Case 7
+            print(clpParser.version)
+            None
+          case ParseResult.Success => // Case 8
+            val clp: Clp = clpParser.instance.get
+            try {
+              clpBlock(clp)
+              Some(clp)
+            }
+            catch {
+              case ex: ValidationException =>
+                printExtraUsage(clazz=Some(clazz))
+                print(clpParser.usage(withPreamble=false))
+                ex.messages.foreach(msg => print(wrapError(msg)))
+                None
+            }
+        }
+    }
+  }
+
+  /** Method for parsing a main classe's command line arguments as well as a specific command line program's arguments.
```

Comment:
I think we should use `command` and `sub-command` to differentiate between the top level and the subordinate programs since that's what every other arg/opt parser calls them.